enhancements/network/admin-network-policy.md

```diff
@@ -0,0 +1,322 @@
+
+## Proposal
+
+A new level-driven controller will be added in OVN-Kubernetes which will
+watch for the `AdminNetworkPolicy`, `BaselineAdminNetworkPolicy`, `Namespaces`
+and `Pods` objects and create the:
+
+* port groups for the subject pods selected in the policies - we will try to
+investigate if shared port groups can be used here whereever possible
+* address-sets for the peers defined in the policies - we will try and reuse
+the address sets that are created for namespaces today and share them wherever possible
+* acls for the ingress and egress policy rules that will connect these port-groups
+and address-sets together into meaningful allow, deny or pass rules
+
+for:
+
+* ANP add/update/delete
+* BANP add/update/delete
+* pod add/update/delete (if it matches the policies; NOTE: updates are handled
+only if pod labels, podIPs or podState (completed) changed)
+* namespace add/update/delete (if it matches the policies; NOTE: updates are
+handled only for label changes)
+
+This controller will live in the `go-controller/pkg/ovn/controller` path
+of OVNKubernetes repo under its own package `admin_network_policy`. Since these
+are new APIs, we cannot reuse the [caches](https://github.com/ovn-org/ovn-kubernetes/blob/68295ec197b8eadd61ff872e293676fa1a68179d/go-controller/pkg/ovn/base_network_controller_policy.go#L123) that are present in network policy code.
+The new controller will handle the implementation independent of the existing network
+policy v1 code. A lot of performance fixes specially around locking have gone into
+network policy v1 code and we don't want to destabilize those parts. Neither do we
+want to add more things into the overloaded `oc` controller.
+
+This controller hopes to be the first step towards breaking out network policy
+construct management into its own set of standard controllers for the newer APIs.
+Note that Network Policy V1 API is core and in-tree and potentially will never be
+deprecated. Thus the existing code for network policy can be maintained as is in
+the `oc` controller if need be. But new features are no longer added to that API.
+So whenever Developer Network Policies come into play which is the future v2 of
+network policy v1, we should be able to leverage the design of this new controller
+and implement that easily. 
+
+### Hierarchical ACLs and Pass Action
+
+In order to implement the smooth co-existence of ANPs, NPs and BANPs in a cluster,
+and ensure we provide clear precedence in the order of [ANP > NP > BANP](https://deploy-preview-54--kubernetes-sigs-network-policy-api.netlify.app/)
+we will use a new RFE from OVN called [Hierarchical ACLs](https://bugzilla.redhat.com/show_bug.cgi?id=2134138).
+
+This will let us define ACLs in stages or tiers. Currently we can support upto
+4 tiers - 0, 1, 2 and 3. We will allot tier1 to ANP ACLs, tier2 to NP ACLs and
```

Comment:
Why not utilize tiers 0, 1, 2 instead instead?  Is there some other ACL user that will take advantage of tier 0 or is it just to error on the side of caution?

Reply:
yeah so OVN team gave us 4 tiers to start with, I want to leave the tier0 as placeholder for some "SuperAdmin" use cases if any :) 5 years from now. If we want more tiers like 4/5/6/7 etc we can ask OVN team to bump the number from 4 to 8, but if we use tier0 now and need it for something in future then we'd need to move stuff over.

Also 0 is the default value for this column during upgrades, so easier to allot the other numbers as separate from the default value.